internal/client/conn.go

```diff
@@ -45,49 +46,60 @@ type UDPConnObserver interface {
 	OnDeallocated(relayedAddr net.Addr)
 }
 
-// UDPConnConfig is a set of configuration params use by NewUDPConn
-type UDPConnConfig struct {
-	Observer    UDPConnObserver
+// ConnConfig is a set of configuration params use by NewUDPConn and NewTCPConn
+type ConnConfig struct {
+	Dialer      transport.Dialer
+	Observer    ConnObserver
 	RelayedAddr net.Addr
 	Integrity   stun.MessageIntegrity
 	Nonce       stun.Nonce
 	Lifetime    time.Duration
 	Log         logging.LeveledLogger
 }
 
-// UDPConn is the implementation of the Conn and PacketConn interfaces for UDP network connections.
-// comatible with net.PacketConn and net.Conn
-type UDPConn struct {
-	obs               UDPConnObserver       // read-only
+type RelayConn interface {
+	CreatePermissions(addrs ...net.Addr) error
+}
+
+type RelayConnContext struct {
```

Comment:
RelayX are used for both TCP and UDP connections. Both UDPConn and TCPConn use the fields in RelayConnContext. In this case, do you prefer the naming of RelayConnImpl?

In client.go, RelayConn was used to generalize between UDPConn and TCPConn. I think we can get rid of this and each client will have a UDPConn and TCPAllocation.

Reply:
I think the confusion stems from the fact that we do not have a `client.UDPAllocation` type.
If we would have it, `client.RelayConnContext` would be more appropriately named `client.Allocation`.

But in fact, the fields of `client.RelayConnContext`are basically just the clients per-allocation state.

Reply:
I am just confused why we have a per-allocation `_nonce` and and `integrity` fields here. Also the naming of `ConnObserver` is confusing.

I think this could need some refactoring as well.
- Rename `RelayConnContext` to `Allocation`
- Rename `ConnObserver` to `Client` (Just an interface within `internal/client` representing the public `Client` type in order to break cyclic dependencies)
- Move integrity and nonce fields from `RelayConnContext`  to the new `Client interface`
- Rename `conn.go` to `udp_conn.go`
- Move common code from `conn.go` to `allocation.go`